Issue with LevelEventGenericPacket

This packet has "floating" NBT whereas the code expects them in a Compound Tag. I've only seen this packet used so far (for what I'm not sure yet) in MCEE 1.14.31. 


```
[12:48:11 DEBUG]: Error occurred whilst decoding packet
com.nukkitx.protocol.bedrock.exception.PacketSerializeException: Error whilst deserializing LevelEventGenericPacket(eventId=2023, tag=null)
        at com.nukkitx.protocol.bedrock.BedrockPacketCodec.tryDecode(BedrockPacketCodec.java:60) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at com.nukkitx.protocol.bedrock.wrapper.BedrockWrapperSerializerV9_10.deserialize(BedrockWrapperSerializerV9_10.java:73) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at com.nukkitx.protocol.bedrock.BedrockSession.onWrappedPacket(BedrockSession.java:276) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at com.nukkitx.protocol.bedrock.BedrockRakNetSessionListener.onEncapsulated(BedrockRakNetSessionListener.java:32) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at com.nukkitx.network.raknet.RakNetSession.onEncapsulatedInternal(RakNetSession.java:307) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at com.nukkitx.network.raknet.RakNetSession.onOrderedReceived(RakNetSession.java:434) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at com.nukkitx.network.raknet.RakNetSession.checkForOrdered(RakNetSession.java:412) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at com.nukkitx.network.raknet.RakNetSession.onRakNetDatagram(RakNetSession.java:405) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at com.nukkitx.network.raknet.RakNetSession.onDatagram(RakNetSession.java:279) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at com.nukkitx.network.raknet.RakNetClient$ClientDatagramHandler.channelRead(RakNetClient.java:153) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1422) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:931) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.channel.epoll.EpollDatagramChannel.read(EpollDatagramChannel.java:688) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.channel.epoll.EpollDatagramChannel.access$100(EpollDatagramChannel.java:57) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.channel.epoll.EpollDatagramChannel$EpollDatagramChannelUnsafe.epollInReady(EpollDatagramChannel.java:507) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:502) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:407) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.util.concurrent.SingleThreadEventExecutor$6.run(SingleThreadEventExecutor.java:1050) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at java.lang.Thread.run(Thread.java:834) ~[?:?]
Caused by: java.lang.ClassCastException: class java.lang.Integer cannot be cast to class com.nukkitx.nbt.NbtMap (java.lang.Integer is in module java.base of loader 'bootstrap'; com.nukkitx.nbt.NbtMap is in unnamed module of loader 'app')
        at com.nukkitx.protocol.bedrock.v361.serializer.LevelEventGenericSerializer_v361.deserialize(LevelEventGenericSerializer_v361.java:24) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at com.nukkitx.protocol.bedrock.v361.serializer.LevelEventGenericSerializer_v361.deserialize(LevelEventGenericSerializer_v361.java:11) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at com.nukkitx.protocol.bedrock.BedrockPacketCodec.tryDecode(BedrockPacketCodec.java:58) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        ... 24 more
[12:48:20 DEBUG]: Error occurred whilst decoding packet
com.nukkitx.protocol.bedrock.exception.PacketSerializeException: Error whilst deserializing LevelEventGenericPacket(eventId=2023, tag=null)
        at com.nukkitx.protocol.bedrock.BedrockPacketCodec.tryDecode(BedrockPacketCodec.java:60) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at com.nukkitx.protocol.bedrock.wrapper.BedrockWrapperSerializerV9_10.deserialize(BedrockWrapperSerializerV9_10.java:73) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at com.nukkitx.protocol.bedrock.BedrockSession.onWrappedPacket(BedrockSession.java:276) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at com.nukkitx.protocol.bedrock.BedrockRakNetSessionListener.onEncapsulated(BedrockRakNetSessionListener.java:32) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at com.nukkitx.network.raknet.RakNetSession.onEncapsulatedInternal(RakNetSession.java:307) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at com.nukkitx.network.raknet.RakNetSession.onOrderedReceived(RakNetSession.java:434) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at com.nukkitx.network.raknet.RakNetSession.checkForOrdered(RakNetSession.java:412) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at com.nukkitx.network.raknet.RakNetSession.onRakNetDatagram(RakNetSession.java:405) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at com.nukkitx.network.raknet.RakNetSession.onDatagram(RakNetSession.java:279) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at com.nukkitx.network.raknet.RakNetClient$ClientDatagramHandler.channelRead(RakNetClient.java:153) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1422) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:931) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.channel.epoll.EpollDatagramChannel.read(EpollDatagramChannel.java:688) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.channel.epoll.EpollDatagramChannel.access$100(EpollDatagramChannel.java:57) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.channel.epoll.EpollDatagramChannel$EpollDatagramChannelUnsafe.epollInReady(EpollDatagramChannel.java:507) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:502) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:407) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.util.concurrent.SingleThreadEventExecutor$6.run(SingleThreadEventExecutor.java:1050) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at java.lang.Thread.run(Thread.java:834) ~[?:?]
Caused by: java.lang.ClassCastException: class java.lang.Integer cannot be cast to class com.nukkitx.nbt.NbtMap (java.lang.Integer is in module java.base of loader 'bootstrap'; com.nukkitx.nbt.NbtMap is in unnamed module of loader 'app')
        at com.nukkitx.protocol.bedrock.v361.serializer.LevelEventGenericSerializer_v361.deserialize(LevelEventGenericSerializer_v361.java:24) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at com.nukkitx.protocol.bedrock.v361.serializer.LevelEventGenericSerializer_v361.deserialize(LevelEventGenericSerializer_v361.java:11) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at com.nukkitx.protocol.bedrock.BedrockPacketCodec.tryDecode(BedrockPacketCodec.java:58) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        ... 24 more
[12:48:21 DEBUG]: Error occurred whilst decoding packet
com.nukkitx.protocol.bedrock.exception.PacketSerializeException: Error whilst deserializing LevelEventGenericPacket(eventId=2023, tag=null)
        at com.nukkitx.protocol.bedrock.BedrockPacketCodec.tryDecode(BedrockPacketCodec.java:60) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at com.nukkitx.protocol.bedrock.wrapper.BedrockWrapperSerializerV9_10.deserialize(BedrockWrapperSerializerV9_10.java:73) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at com.nukkitx.protocol.bedrock.BedrockSession.onWrappedPacket(BedrockSession.java:276) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at com.nukkitx.protocol.bedrock.BedrockRakNetSessionListener.onEncapsulated(BedrockRakNetSessionListener.java:32) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at com.nukkitx.network.raknet.RakNetSession.onEncapsulatedInternal(RakNetSession.java:307) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at com.nukkitx.network.raknet.RakNetSession.onOrderedReceived(RakNetSession.java:434) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at com.nukkitx.network.raknet.RakNetSession.checkForOrdered(RakNetSession.java:412) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at com.nukkitx.network.raknet.RakNetSession.onRakNetDatagram(RakNetSession.java:405) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at com.nukkitx.network.raknet.RakNetSession.onDatagram(RakNetSession.java:279) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at com.nukkitx.network.raknet.RakNetClient$ClientDatagramHandler.channelRead(RakNetClient.java:153) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1422) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:931) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.channel.epoll.EpollDatagramChannel.read(EpollDatagramChannel.java:688) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.channel.epoll.EpollDatagramChannel.access$100(EpollDatagramChannel.java:57) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.channel.epoll.EpollDatagramChannel$EpollDatagramChannelUnsafe.epollInReady(EpollDatagramChannel.java:507) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.channel.epoll.AbstractEpollChannel$AbstractEpollUnsafe$1.run(AbstractEpollChannel.java:394) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:510) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:413) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.util.concurrent.SingleThreadEventExecutor$6.run(SingleThreadEventExecutor.java:1050) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at java.lang.Thread.run(Thread.java:834) ~[?:?]
Caused by: java.lang.ClassCastException: class java.lang.Integer cannot be cast to class com.nukkitx.nbt.NbtMap (java.lang.Integer is in module java.base of loader 'bootstrap'; com.nukkitx.nbt.NbtMap is in unnamed module of loader 'app')
        at com.nukkitx.protocol.bedrock.v361.serializer.LevelEventGenericSerializer_v361.deserialize(LevelEventGenericSerializer_v361.java:24) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at com.nukkitx.protocol.bedrock.v361.serializer.LevelEventGenericSerializer_v361.deserialize(LevelEventGenericSerializer_v361.java:11) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        at com.nukkitx.protocol.bedrock.BedrockPacketCodec.tryDecode(BedrockPacketCodec.java:58) ~[proxypass-1.0.0-SNAPSHOT.jar:?]
        ... 26 more
```

--- a/bedrock/bedrock-common/src/main/java/com/nukkitx/protocol/bedrock/BedrockPacketHelper.java
+++ b/bedrock/bedrock-common/src/main/java/com/nukkitx/protocol/bedrock/BedrockPacketHelper.java
@@ -5,6 +5,7 @@ import com.nukkitx.math.vector.Vector3f;
 import com.nukkitx.math.vector.Vector3i;
 import com.nukkitx.nbt.NBTInputStream;
 import com.nukkitx.nbt.NBTOutputStream;
+import com.nukkitx.nbt.NbtType;
 import com.nukkitx.nbt.NbtUtils;
 import com.nukkitx.network.VarInts;
 import com.nukkitx.network.util.Preconditions;
@@ -547,6 +548,22 @@ public abstract class BedrockPacketHelper {
     public <T> void writeTag(ByteBuf buffer, T tag) {
         try (NBTOutputStream writer = NbtUtils.createNetworkWriter(new ByteBufOutputStream(buffer))) {
             writer.writeTag(tag);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public <T> T readTagValue(ByteBuf buffer, NbtType<T> type) {
+        try (NBTInputStream reader = NbtUtils.createNetworkReader(new ByteBufInputStream(buffer))) {
+            return reader.readValue(type);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public <T> void writeTagValue(ByteBuf buffer, T tag) {
+        try (NBTOutputStream writer = NbtUtils.createNetworkWriter(new ByteBufOutputStream(buffer))) {
+            writer.writeValue(tag);
         } catch (IOException e) {
             throw new RuntimeException(e);
         }

--- a/bedrock/bedrock-v361/src/main/java/com/nukkitx/protocol/bedrock/v361/serializer/LevelEventGenericSerializer_v361.java
+++ b/bedrock/bedrock-v361/src/main/java/com/nukkitx/protocol/bedrock/v361/serializer/LevelEventGenericSerializer_v361.java
@@ -1,5 +1,6 @@
 package com.nukkitx.protocol.bedrock.v361.serializer;
 
+import com.nukkitx.nbt.NbtType;
 import com.nukkitx.network.VarInts;
 import com.nukkitx.protocol.bedrock.BedrockPacketHelper;
 import com.nukkitx.protocol.bedrock.BedrockPacketSerializer;
@@ -15,12 +16,12 @@ public class LevelEventGenericSerializer_v361 implements BedrockPacketSerializer
     @Override
     public void serialize(ByteBuf buffer, BedrockPacketHelper helper, LevelEventGenericPacket packet) {
         VarInts.writeInt(buffer, packet.getEventId());
-        helper.writeTag(buffer, packet.getTag());
+        helper.writeTagValue(buffer, packet.getTag());
     }
 
     @Override
     public void deserialize(ByteBuf buffer, BedrockPacketHelper helper, LevelEventGenericPacket packet) {
         packet.setEventId(VarInts.readInt(buffer));
-        packet.setTag(helper.readTag(buffer));
+        packet.setTag(helper.readTagValue(buffer, NbtType.COMPOUND));
     }
 }

--- a/common/pom.xml
+++ b/common/pom.xml
@@ -27,7 +27,7 @@
         <dependency>
             <groupId>com.nukkitx</groupId>
             <artifactId>nbt</artifactId>
-            <version>2.0.1</version>
+            <version>2.2.0</version>
             <scope>compile</scope>
         </dependency>
         <dependency>

